src/logic/synchronization.ts

```diff
@@ -23,46 +25,63 @@ import {
-            .map(type => ({
-                type,
-                ids: references[type],
-                excludeRules: nestedExcludeRules[type],
-                includeRules: nestedIncludeRules[type],
-            }))
-            .map(newBuilder => exportMetadata(d2, newBuilder));
-        const promisesResult: MetadataPackage[] = await Promise.all(promises);
-        _.deepMerge(result, ...promisesResult);
-    }
-
-    // Clean up result from duplicated elements
-    _.forOwn(result, (value, metadataType) => (result[metadataType] = _.uniqBy(value, "id")));
-    return result;
+async function exportMetadata(d2: D2, originalBuilder: ExportBuilder): Promise<MetadataPackage> {
+    const visitedIds: Set<string> = new Set();
+    const recursiveExport = async (builder: ExportBuilder): Promise<MetadataPackage> => {
+        const { type, ids, excludeRules, includeRules } = builder;
+        const model = d2ModelFactory(d2, type).getD2Model(d2);
+        const result: MetadataPackage = {};
+
+        // Each level of recursion traverse the exclude/include rules with nested values
+        const nestedExcludeRules: NestedRules = buildNestedRules(excludeRules);
+        const nestedIncludeRules: NestedRules = buildNestedRules(includeRules);
+
+        // Get all the required metadata
+        const syncMetadata = await getMetadata(d2, ids);
+        const elements = syncMetadata[model.plural] || [];
+
+        for (const element of elements) {
+            // Store metadata object in result
+            const object = cleanObject(element, excludeRules);
+            result[model.plural] = result[model.plural] || [];
+            result[model.plural].push(object);
+
+            // Get all the referenced metadata
+            const references: MetadataPackage = getAllReferences(d2, object, model.name);
+            const includedReferences = cleanReferences(references, includeRules);
+            const promises = includedReferences
+                .map(
+                    (type: string): ExportBuilder => ({
+                        type,
+                        ids: references[type].filter((id: string): boolean => !visitedIds.has(id)),
+                        excludeRules: nestedExcludeRules[type],
+                        includeRules: nestedIncludeRules[type],
+                    })
+                )
+                .map(
+                    (newBuilder: ExportBuilder): Promise<MetadataPackage> => {
+                        newBuilder.ids.forEach(
+                            (id: string): void => {
+                                visitedIds.add(id);
+                            }
+                        );
+                        return recursiveExport(newBuilder);
+                    }
+                );
+            const promisesResult: MetadataPackage[] = await Promise.all(promises);
+            _.deepMerge(result, ...promisesResult);
+        }
+
+        // Clean up result from duplicated elements
+        _.forOwn(
```

Comment:
`result` needs to be updated imperatively? Functional approach:

return _.mapValues(result, (objects, metadataType) => _.uniqBy(objects, "id"))